rust_src/src/keymap.rs

```diff
@@ -62,6 +65,193 @@ pub extern "C" fn set_where_is_cache(val: LispObject) {
     }
 }
 
+// Start here **************************
+
+// TODO Change this wherever it is used in rust to this implementation
+// TODO: Is this function used anywhere? ripgrep cannot seem to find it
+
+// This function has an extra argument called dummy but it is not used.
+// original signature is :
+// map_keymap_call (Lisp_Object key, Lisp_Object val, Lisp_Object fun, void *dummy)
+// For now, this is simply ported and the dummy argument is ommitted
+#[no_mangle]
+pub extern "C" fn _map_keymap_call(key: LispObject, val: LispObject, fun: LispObject) {
+    unsafe {
+        call2(fun, key, val);
+    }
+}
+
+// TODO Change this wherever it is used in rust to this implementation
+//
+/// Same as map_keymap, but does it right, properly eliminating duplicate
+/// bindings due to inheritance.
+#[no_mangle]
+pub extern "C" fn _map_keymap_canonical(
+    map: LispObject,
+    fun: map_keymap_function_t,
+    args: LispObject,
+    data: *mut c_void,
+) {
+    unsafe {
+        let map = safe_call1(Qkeymap_canonicalize, map);
+        map_keymap_internal(map, fun, args, data);
+    }
+}
+
+// Scopes are all messed up. Double check those
+
+/* Given OBJECT which was found in a slot in a keymap,
+trace indirect definitions to get the actual definition of that slot.
+An indirect definition is a list of the form
+(KEYMAP . INDEX), where KEYMAP is a keymap or a symbol defined as one
+and INDEX is the object to look up in KEYMAP to yield the definition.
+
+Also if OBJECT has a menu string as the first element,
+remove that.  Also remove a menu help string as second element.
+
+If AUTOLOAD, load autoloadable keymaps
+that are referred to with indirection.
+
+This can GC because menu_item_eval_property calls Feval.  */
+#[no_mangle]
+pub extern "C" fn _get_keyelt(object: LispObject, autoload: bool) -> LispObject {
+    loop {
```

Comment:
```suggestion
    while let Some(cons) = object.as_cons() {
```

Does that read better? Then if the "if not cons" and "as_cons" below are not needed.

Reply:
Actually, reading below, it appears we could go a little further and instead do:

`while let Some(car, cdr) = object.into() {`

However, I would suggest we find a better name for `car` and `cdr` here.